torch/fx/passes/reinplace.py

```diff
@@ -0,0 +1,540 @@
+
+    Sample inputs are needed to determine aliasing relationships of the inputs.
+    In general, we can't reinplace node `b = a.add(...)` if "a" aliases any of the
+    inputs to the program.
+
+    Given a node "b = foo(a, ...)", the algorithm for re-inplacing is as follows:
+
+    (1) Check if foo has a mutating variant. If not, move to the next node.
+
+        Note that we ignore view ops (we don't bother to turn `as_strided()`
+        into `as_strided_()`), as it complicates the algorithm and doesn't
+        provide meaningful speedups.
+
+        Currently, we also only check for an inplace op, `foo_`.
+        Later, we should beef this up to check for out= or mutable ops.
+
+    (2) Check if "a" is an alias of any of the program inputs.
+
+        If it is, skip and move to the next node.
+        Inplace'ing an op that would cause it to mutate a program is not sound,
+        because that would be a side effect visible to the user.
+
+        NOTE: there's a future optimization that we should make:
+        if "a" is a (alias of a)  program input, but later in the program
+        there is a node that looks like "a.copy_(...)",
+        Then re-inplacing is ok to do - we are temporarily re-using a's buffer,
+        which will later be overwritten by the copy_() call.
+
+        This will be an important optimization to have for programs that mutate
+        their inputs. It currently isn't implemented though.
+
+    (3) Check that "a" and all of its outstanding aliases are not used anywhere
+        later in the graph. If this is the case, then it's safe to re-inplace
+        to "b = foo_(a)".
+
+        There are a few caveats to this, explained in more detail below:
+        (a) If "a" is used later as an argument to a view op, that is okay.
+            It's only a problem if "a" (or that view) is later passed
+            into a normal operator, or if it is returned as the program output.
+        (b) If "a" is used as an input into a view "inverse" / "scatter"
+            operator, it is potentially fine to re-inplace
+            (and remove that scatter operator from the graph).
+            See below for a more detailed example.
+
+        NOTE: there is an optimization in this step that is crucial
+        to fully recovering performance from functionalization.
+
+        Given this program:
+        def f(x):
+            a = torch.ops.aten.add(x, x)
+            b = torch.ops.aten.diagonal(a)
+            torch.ops.aten.fill_(b, 0)
+            return d
+
+        Functionalization will emit the following:
+        def f(x):
+            a = torch.ops.aten.add(x, x)
+            b = torch.ops.aten.diagonal(a, 0, 1)
+            b_updated = torch.ops.aten.fill(b, 0)
+            a_updated = torch.ops.aten.diagonal_scatter(a, b_updated, 0, 1)
+            return a_updated
+
+        And "re-inplacing" is on the hook for figuring out how to remove
+        the expensive diagonal_scatter call.
```

Comment:
"Ordinarily, we would not be able to reinplace the fill, because b aliases with a which is used by the diagonal_scatter call"